test: use MiniWallet to simplify mempool_package_limits.py tests

kouloumos opened the pull request:
While `wallet.py` includes the MiniWallet class and some helper methods, it also includes some methods that have been moved there without having any direct relation with the MiniWallet class. Specifically `make_chain`, `create_child_with_parents` and `create_raw_chain` methods that were extracted from `rpc_packages.py` at f8253d69d6f02850995a11eeb71fedc22e6f6575 in order to be used on both `mempool_package_limits.py` and `rpc_packages.py`.

Since that change, due to the introduction of additional methods in MiniWallet, the functionality of those methods can now be replicated with the existing MiniWallet methods and simultaneously simplify those tests by using the MiniWallet.

This PR's goals are

-  to simplify the `mempool_package_limits.py` functional tests with usage of the MiniWallet.
-  to make progress towards the removal of the `make_chain`, `create_child_with_parents` and `create_raw_chain` methods of `wallet.py`.

For the purpose of the aforementioned goals, a helper method `MiniWallet.send_self_transfer_chain` is introduced and method `bulk_transaction` has been integrated in `create_self_transfer*` methods using an optional `target_weight` option.

kouloumos committed:
test: use MiniWallet to simplify mempool_package_limits.py tests

Moved `bulk_transaction` into MiniWallet class as `_bulk_tx` private
helper method to be used when the newly added `target_weight` option is
passed to `create_self_transfer*`

MarcoFalke commented on test/functional/mempool_package_limits.py:
```diff
@@ -6,48 +6,30 @@
+        # MiniWallet's default address in blocks 76-100 (see method
+        # BitcoinTestFramework._initialize_chain())
+        self.wallet.rescan_utxos()
+        self.generate(self.wallet, 100) # more mature UTXOs are needed
```
I am pretty sure none of the utxos are mature, so the comment doesn't apply.

so maybe this can be removed?

If not, you may just mine 200 blocks in one go instead of using cached/non-cached?

kouloumos replied:
It matures the cached blocks, so I could phrase that in better way. But you are right, I changed it to mine the neccesary blocks on a clean chain.

MarcoFalke replied:
blocks 76-100 with a height of 200 should already be mature, no?

kouloumos replied:
Yes but because those 25 blocks were not enough for the tests, I also used the 24 initially immature blocks by generating another 100 blocks. But what you suggested is more elegant thus I changed it to that.

kouloumos committed:
test: MiniWallet: add `send_self_transfer_chain` to create chain of txns

With this new method, a chain of transactions can be created. This
method is introduced to further simplify the mempool_package_limits.py
tests.

kouloumos commented:
Forced-pushed to take advantage of the logic introduced in https://github.com/bitcoin/bitcoin/pull/25445

MarcoFalke commented on test/functional/test_framework/wallet.py:
```diff
@@ -333,6 +333,19 @@ def bulk_transaction(self, tx, target_weight):
             "tx": tx_heavy,
         }
 
+    def send_self_transfer_chain(self, *, from_node=None, chain_length, utxo_to_spend=None):
```
1c9243b40db5746575ed55d09774872f0a98b24a: `from_node` should be a required arg, like in the other `send*` methods.

maflcko reviewed (approved):
lgtm 1c9243b40db5746575ed55d09774872f0a98b24a 🐐

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

lgtm 1c9243b40db5746575ed55d09774872f0a98b24a 🐐
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjMzgv8Ck6nSSmgAatlxCMG837PArkz0UTGPh4ek3WbN73Mn1dLmSCBsvlzbSVG
4yQlx+vMpDNUrJO07nU7fVszAKK5+BOFUcd1pElpZbIoOWHIudbDdEuOeOIDjuKC
368yHXtcXZUnx4ZUGJvC11LtkH2eZxjrwt70iVWqoNWMz2Wj3dNS01VGOkAN6Bb2
8wUWoNYwgm+iTNJIsQnzPuc9I7qjhILvhcbMTXKX/NQQi4StPq+5nBRFIXSePSuD
ApxcDIGIB2s3cYu/SswObSHtCtRKt/gaG0Xd7OkOMZ8bpNPU0vWiRAFtRn5OhKj7
2csGoyU1prVU52njdCYaC/EAiyWumqBmw/4CtffajgZ4UJluPhvn2ZBjAJDq9DAb
SiFfaITnyqEA327CahBPIkyTXmL15HC0zipUDjs0U5YYLby2z05h6i9htpQv+iCJ
eH1N+nmixN9aA6v1gkXNEs7qVcVWWLhLUuatmxzgaEr5pdhn+7701yNirhqu004l
JrbTuRmo
=dEET
-----END PGP SIGNATURE-----
```


</details>

neto626roman commented:
Eject ur project

On Mon, Aug 1, 2022, 5:08 AM MacroFake ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> lgtm 1c9243b
> <https://github.com/bitcoin/bitcoin/commit/1c9243b40db5746575ed55d09774872f0a98b24a>
> 🐐
> Show signature
>
> Signature:
>
> -----BEGIN PGP SIGNED MESSAGE-----
>
> Hash: SHA512
>
>
>
> lgtm 1c9243b40db5746575ed55d09774872f0a98b24a 🐐
>
> -----BEGIN PGP SIGNATURE-----
>
>
>
> iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
>
> pUjMzgv8Ck6nSSmgAatlxCMG837PArkz0UTGPh4ek3WbN73Mn1dLmSCBsvlzbSVG
>
> 4yQlx+vMpDNUrJO07nU7fVszAKK5+BOFUcd1pElpZbIoOWHIudbDdEuOeOIDjuKC
>
> 368yHXtcXZUnx4ZUGJvC11LtkH2eZxjrwt70iVWqoNWMz2Wj3dNS01VGOkAN6Bb2
>
> 8wUWoNYwgm+iTNJIsQnzPuc9I7qjhILvhcbMTXKX/NQQi4StPq+5nBRFIXSePSuD
>
> ApxcDIGIB2s3cYu/SswObSHtCtRKt/gaG0Xd7OkOMZ8bpNPU0vWiRAFtRn5OhKj7
>
> 2csGoyU1prVU52njdCYaC/EAiyWumqBmw/4CtffajgZ4UJluPhvn2ZBjAJDq9DAb
>
> SiFfaITnyqEA327CahBPIkyTXmL15HC0zipUDjs0U5YYLby2z05h6i9htpQv+iCJ
>
> eH1N+nmixN9aA6v1gkXNEs7qVcVWWLhLUuatmxzgaEr5pdhn+7701yNirhqu004l
>
> JrbTuRmo
>
> =dEET
>
> -----END PGP SIGNATURE-----
>
>
> ------------------------------
>
> In test/functional/mempool_package_limits.py
> <https://github.com/bitcoin/bitcoin/pull/25379#discussion_r934441857>:
>
> > @@ -486,50 +369,21 @@ def test_desc_size_limits(self):
>
>          high_fee = Decimal("0.0021") # 10 sats/vB
>
>
> 52424d2
> <https://github.com/bitcoin/bitcoin/commit/52424d27e0990cc3b77ca5d6e45f045cbe1eaa60>:
> This is wrong. Now it is 20sat/vB
> ------------------------------
>
> In test/functional/test_framework/wallet.py
> <https://github.com/bitcoin/bitcoin/pull/25379#discussion_r934442790>:
>
> > +            for _ in range(512):
>
> +                script_pubkey = script_pubkey + "01"
>
> +            tx_heavy.vout.append(CTxOut(0, bytes.fromhex(script_pubkey)))
>
> +
>
> +        txid = tx_heavy.rehash()
>
> +        return {
>
> +            "new_utxos": [self._create_utxo(
>
> +                txid=txid,
>
> +                vout=i,
>
> +                value=Decimal(tx.vout[i].nValue) / COIN,
>
> +                height=0,
>
> +            ) for i in range(len(tx.vout))],
>
> +            "txid": txid,
>
> +            "hex": tx_heavy.serialize().hex(),
>
> +            "tx": tx_heavy,
>
> +        }
>
>
> 52424d2
> <https://github.com/bitcoin/bitcoin/commit/52424d27e0990cc3b77ca5d6e45f045cbe1eaa60>:
> Any reason to not make this a target_weight=0 option on create_self_tx to
> avoid all the duplicated logic?
> ------------------------------
>
> In test/functional/mempool_package_limits.py
> <https://github.com/bitcoin/bitcoin/pull/25379#discussion_r934448307>:
>
> >  from test_framework.util import (
>
>      assert_equal,
>
>  )
>
> -from test_framework.wallet import (
>
> -    bulk_transaction,
>
>
> Can remove all of these functions now?
> ------------------------------
>
> In test/functional/test_framework/wallet.py
> <https://github.com/bitcoin/bitcoin/pull/25379#discussion_r934450695>:
>
> > @@ -333,6 +333,19 @@ def bulk_transaction(self, tx, target_weight):
>
>              "tx": tx_heavy,
>
>          }
>
>
>
> +    def send_self_transfer_chain(self, *, from_node=None, chain_length, utxo_to_spend=None):
>
>
> 1c9243b
> <https://github.com/bitcoin/bitcoin/commit/1c9243b40db5746575ed55d09774872f0a98b24a>:
> from_node should be a required arg, like in the other send* methods.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/25379#pullrequestreview-1057172168>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARWXJNWGLROFTIUT3W2UQTDVW644RANCNFSM5Y3RGJJA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

MarcoFalke commented:
Also, should be rebased

kouloumos commented on test/functional/mempool_package_limits.py:
```diff
@@ -486,50 +369,21 @@ def test_desc_size_limits(self):
         high_fee = Decimal("0.0021") # 10 sats/vB
```
I've now fixed that with a division on the method call.

kouloumos commented on test/functional/test_framework/wallet.py:
```diff
@@ -309,6 +308,30 @@ def sendrawtransaction(self, *, from_node, tx_hex, maxfeerate=0, **kwargs):
+            "txid": txid,
+            "hex": tx_heavy.serialize().hex(),
+            "tx": tx_heavy,
+        }
```
Good idea. I've now created `_bulk_tx` because the logic is both in `create_self_transfer` and `create_self_transfer_multi`

kouloumos reviewed (commented):
Rebased and addressed @MarcoFalke comments.

~The failing CI task is also on master, it seems not to be related with the changes.~

MarcoFalke commented on test/functional/test_framework/wallet.py:
```diff
@@ -104,6 +103,17 @@ def __init__(self, test_node, *, mode=MiniWalletMode.ADDRESS_OP_TRUE):
+        """
+        assert_greater_than_or_equal(target_weight, tx.get_weight())
+        while tx.get_weight() < target_weight:
+            script_pubkey = "6a4d0200"  # OP_RETURN OP_PUSH2 512 bytes
+            for _ in range(512):
+                script_pubkey = script_pubkey + "01"
+            tx.vout.append(CTxOut(0, bytes.fromhex(script_pubkey)))
```
```suggestion
            script_pubkey = ( b"6a4d0200"  # OP_RETURN OP_PUSH2 512 bytes
             + b"01" * 512 )
            tx.vout.append(CTxOut(0, script_pubkey))
```

nit?

kouloumos replied:
Done

MarcoFalke commented:
ACK f2f6068b69b1b532db92b276f024c89b56f38294 👜

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK f2f6068b69b1b532db92b276f024c89b56f38294 👜
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjJEAwAs/RUpTTagS6DLfFVOT2d+FAmEcS+/m1s5438bxGA9pb/lhVYxxYlOTvM
wtgbP2N1G496b2wmgzRWZ8LPdI4JvUmzgP4+f9fvPO2jXo6+bLZFfLBnVvhbybVn
keiIkU38nUmohPlQ2OZvNfXFxrf3Ij1jQDauEcg2wk4fjog7WPFK3tkA6Js338bE
/GzH7hmy/pNMMTZcsX6etEwxT+zOfit/PKo4b7+VD1R41lkSg1SacUdMbUph13Lw
bVtPG450XzW0U2MSS8iPyG81y1Wnlcvtj2Z8wO5MsSuJbcoYT5/haDjsMssKDHF9
YpRhKbRVYeejJny9xIqrtGP56WIjIdGrJ7h3I14vmn9zqeu4qvW9nUeUN7VHyEiw
dTb8YDFwTjmI6h2XzVA2OBltMB4ZlUJqvL59CO6DOHKwzObtWh/Y5ox2GKqEzEUz
QXBwKrNA3g+5jKAJYfpRy6c8OTTsN4elrnl45T5QmVXe5npyStD3TMPoZJhCASgI
3a/nCGHb
=PXPl
-----END PGP SIGNATURE-----
```


</details>

glozow reviewed (commented):
Concept ACK

I was just thinking it would be nice to get rid of these functions that are only used in package functional tests and use MiniWallet instead, thanks for working on this!

theStack reviewed (commented):
Post-merge concept and light code-review ACK f2f6068b69b1b532db92b276f024c89b56f38294 🌇
